backend/src/models/reports/Tab_41_rpt_PF_PortfolioStaffRecoveries.js

```diff
@@ -1,16 +1,135 @@
+            RIGHT JOIN project_budget pb ON po.id = pb.recovery_area
+          ) ON pd.id = pb.project_deliverable_id
+        ) ON p.id = pd.project_id`
+      )
+      .whereRaw(`left(pb.stob, 2) = '88'`)
+      .groupBy(
+        "po.id",
+        "po.portfolio_name",
+        "p.project_number",
+        "p.project_name",
+        "fy.fiscal_year",
+        "fy.id",
+        "pb.recovery_area",
+        "pb.stob",
+        "p.portfolio_id",
+        "po.portfolio_abbrev"
+      )
+      .orderBy("portfolio_name", "asc")
+      .orderBy("project_number", "asc")
+      .orderBy("fiscal_year", "desc")
+      .modify(whereInArray, "fy.id", fiscal)
+      .modify(whereInArray, "po.id", portfolio),
+
+  // The query to get the report totals.
+  totals: (portfolio, fiscal) =>
+    knex(queries.report(portfolio, fiscal).as("report"))
+      .select({
+        portfolio_name: "portfolio_name",
+      })
+      .sum(queries.columns)
+      .groupBy("portfolio_name"),
+
+  // The query to get the report grand totals.
+  grandTotals: (portfolio, fiscal) =>
+    knex(queries.report(portfolio, fiscal).as("report")).sum(queries.columns).first(),
+};
+
+/**
+ * Retrieve and process data from queries to create a structured result object.
+ *
+ * @param   {object}                  options           - Options object containing fiscal year.
+ * @param   {string | Array<string> } options.portfolio - The portfolio grab data for.
+ * @param   {number}                  options.fiscal    - The fiscal year to grab data for.
+ * @returns {object}                                    - An object containing fiscal year, report, and report total.
+ */
+const getAll = async ({ portfolio, fiscal }) => {
+  try {
+    const reportData = await getReportData(portfolio, fiscal);
+    const reportWithSubtotals = await getReportBySectionWithSubtotalsFrom(
+      reportData,
+      "portfolio_name"
+    );
+
+    return { report: reportWithSubtotals, grand_totals: reportData.grand_totals };
+  } catch (error) {
+    handleGetAllError(error);
+  }
+};
+
+// Execute queries to get the report, totals, and grand totals
+const getReportData = async (portfolio, fiscal) => {
+  const [report, totals, grand_totals] = await Promise.all([
+    queries.report(portfolio, fiscal),
+    queries.totals(portfolio, fiscal),
+    queries.grandTotals(portfolio, fiscal),
+  ]);
+
+  return { report, totals, grand_totals };
+};
+
+// Get report with subtotals folded into each section of the report
+const getReportBySectionWithSubtotalsFrom = async (reportData, sectionName) => {
+  const { report, totals } = reportData;
+  const reportWithSubtotals = await getReportWithSubtotals(report, totals, sectionName);
+
+  return reportWithSubtotals;
+};
+
+// Handle the error thrown by the getAll function.
+const handleGetAllError = (error) => {
+  log.error(error);
+  throw new Error("Error retrieving data for Portfolio Staff Recoveries Report.");
 };
+
+// Exports
+module.exports = { required: [], getAll };
```

Comment:
What's the purpose of exporting an empty array as `required` ?

Reply:
in this model there are zero required parameters

that means that the report could narrow down the results by only grabbing rows where there is a specific portfolio, or a specific fiscal year

however, in this case, that's not mandatory, and the user can run a report for ALL years and ALL portfolios

Reply:
the other models will have something like ["fiscal"]

and in the `genericReportController.js`, we send the option selected by the user on the front-end dropdown menus. They can leave the fiscal year and portfolio blank in this report, and it will still print a (very long) report

Reply:
Oh okay, so some models require parameters to be used. And this way of indicating that is idiomatic to the PMO app. Thanks for explaining!